File: lib/reader.cc
#include "globals.h"
#include "flags.h"
#include "usb/usb.h"
#include "fluxsource/fluxsource.h"
#include "fluxsink/fluxsink.h"
#include "reader.h"
#include "fluxmap.h"
#include "decoders/decoders.h"
#include "sector.h"
#include "bytes.h"
#include "decoders/rawbits.h"
#include "flux.h"
#include "image.h"
#include "imagewriter/imagewriter.h"
#include "logger.h"
#include "fmt/format.h"
#include "proto.h"
#include "utils.h"
#include "mapper.h"
#include "lib/decoders/decoders.pb.h"
#include <iostream>
#include <fstream>

static std::unique_ptr<FluxSink> outputFluxSink;

static std::shared_ptr<Fluxmap> readFluxmap(FluxSource& fluxsource, unsigned cylinder, unsigned head)
{
	Logger() << BeginReadOperationLogMessage { cylinder, head };
	std::shared_ptr<Fluxmap> fluxmap = fluxsource.readFlux(cylinder, head);
	fluxmap->rescale(1.0/config.flux_source().rescale());
	Logger() << EndReadOperationLogMessage()
			 << fmt::format("{0:.0} ms in {1} bytes", fluxmap->duration()/1e6, fluxmap->bytes());
	return fluxmap;
}

static bool conflictable(Sector::Status status)
{
    return (status == Sector::OK) || (status == Sector::CONFLICT);
}

static std::set<std::shared_ptr<const Sector>> collect_sectors(
    std::set<std::shared_ptr<const Sector>>& track_sectors)
{
    typedef std::tuple<unsigned, unsigned, unsigned> key_t;
    std::multimap<key_t, std::shared_ptr<const Sector>> sectors;

    for (const auto& sector : track_sectors)
    {
        key_t sectorid = {
            sector->logicalTrack, sector->logicalSide, sector->logicalSector};
        sectors.insert({sectorid, sector});
    }

    std::set<std::shared_ptr<const Sector>> sector_set;
    auto it = sectors.begin();
    while (it != sectors.end())
    {
        auto ub = sectors.upper_bound(it->first);
        auto new_sector = std::accumulate(it,
            ub,
            it->second,
            [&](auto left, auto& rightit) -> std::shared_ptr<const Sector>
            {
                auto& right = rightit.second;
                if ((left->status == Sector::OK) &&
                    (right->status == Sector::OK) &&
                    (left->data != right->data))
                {
                    auto s = std::make_shared<Sector>(*left);
                    s->status = Sector::CONFLICT;
                    return s;
                }
                if (left->status == Sector::CONFLICT)
                    return left;
                if (right->status == Sector::CONFLICT)
                    return right;
                if (left->status == Sector::OK)
                    return left;
                if (right->status == Sector::OK)
                    return right;
                return left;
            });
        sector_set.insert(new_sector);
        it = ub;
    }
    return sector_set;
}

std::shared_ptr<const DiskFlux> readDiskCommand(FluxSource& fluxsource, AbstractDecoder& decoder)
{
    if (config.decoder().has_copy_flux_to())
        outputFluxSink = FluxSink::create(config.decoder().copy_flux_to());

    auto diskflux = std::make_shared<DiskFlux>();
    bool failures = false;
    for (int cylinder : iterate(config.cylinders()))
    {
        for (int head : iterate(config.heads()))
        {
			testForEmergencyStop();

            auto track = std::make_shared<TrackFlux>();
			track->physicalCylinder = cylinder;
			track->physicalHead = head;
			diskflux->tracks.push_back(track);

            std::set<std::shared_ptr<const Sector>> track_sectors;
            std::set<std::shared_ptr<const Record>> track_records;
            Fluxmap totalFlux;

            for (int retry = config.decoder().retries(); retry >= 0; retry--)
            {
                auto fluxmap = readFluxmap(fluxsource, cylinder, head);
                totalFlux.appendDesync().appendBytes(fluxmap->rawBytes());

                auto trackdataflux =
                    decoder.decodeToSectors(fluxmap, cylinder, head);
                track->trackDatas.push_back(trackdataflux);

                track_sectors.insert(trackdataflux->sectors.begin(),
                    trackdataflux->sectors.end());
                track_records.insert(trackdataflux->records.begin(),
                    trackdataflux->records.end());

                bool hasBadSectors = false;
                std::set<unsigned> required_sectors =
                    decoder.requiredSectors(cylinder, head);
                std::set<std::shared_ptr<const Sector>> result_sectors;
                for (const auto& sector : collect_sectors(track_sectors))
                {
                    result_sectors.insert(sector);
                    required_sectors.erase(sector->logicalSector);

                    if (sector->status != Sector::OK)
                        hasBadSectors = true;
                }
                for (unsigned logical_sector : required_sectors)
                {
                    auto sector = std::make_shared<Sector>();
                    sector->logicalSector = logical_sector;
                    sector->status = Sector::MISSING;
                    result_sectors.insert(sector);

                    hasBadSectors = true;
                }

				track->sectors = collect_sectors(result_sectors);

				/* track can't be modified below this point. */
                Logger() << TrackReadLogMessage { track };

                if (hasBadSectors)
                    failures = false;

                if (!hasBadSectors)
                    break;

                if (!fluxsource.retryable())
                    break;
                if (retry == 0)
                    Logger() << fmt::format("giving up");
                else
                    Logger()
                        << fmt::format("retrying; {} retries remaining", retry);
            }

            if (outputFluxSink)
                outputFluxSink->writeFlux(cylinder, head, totalFlux);

            if (config.decoder().dump_records())
            {
                std::cout << "\nRaw (undecoded) records follow:\n\n";
                for (const auto& record : track_records)
                {
                    std::cout << fmt::format("I+{:.2f}us with {:.2f}us clock\n",
                        record->startTime / 1000.0,
                        record->clock / 1000.0);
                    hexdump(std::cout, record->rawData);
                    std::cout << std::endl;
                }
            }

            if (config.decoder().dump_sectors())
            {
                std::cout << "\nDecoded sectors follow:\n\n";
                for (const auto& sector : track_sectors)
                {
                    std::cout << fmt::format(
                        "{}.{:02}.{:02}: I+{:.2f}us with {:.2f}us clock: "
                        "status {}\n",
                        sector->logicalTrack,
                        sector->logicalSide,
                        sector->logicalSector,
                        sector->headerStartTime / 1000.0,
                        sector->clock / 1000.0,
                        Sector::statusToString(sector->status));
                    hexdump(std::cout, sector->data);
                    std::cout << std::endl;
                }
            }
		}
    }

	if (failures)
		Logger() << "Warning: some sectors could not be decoded.";

	std::set<std::shared_ptr<const Sector>> all_sectors;
	for (auto& track : diskflux->tracks)
		for (auto& sector : track->sectors)
			all_sectors.insert(sector);
	all_sectors = collect_sectors(all_sectors);
	diskflux->image = std::make_shared<Image>(all_sectors);

	if (config.has_sector_mapping())
		diskflux->image = std::move(Mapper::remapPhysicalToLogical(*diskflux->image, config.sector_mapping()));

	/* diskflux can't be modified below this point. */
	Logger() << DiskReadLogMessage { diskflux };
	return diskflux;
}

void readDiskCommand(FluxSource& fluxsource, AbstractDecoder& decoder, ImageWriter& writer)
{
	auto diskflux = readDiskCommand(fluxsource, decoder);

    writer.printMap(*diskflux->image);
    if (config.decoder().has_write_csv_to())
        writer.writeCsv(*diskflux->image, config.decoder().write_csv_to());
    writer.writeImage(*diskflux->image);
}

void rawReadDiskCommand(FluxSource& fluxsource, FluxSink& fluxsink)
{
	for (int cylinder : iterate(config.cylinders()))
	{
		for (int head : iterate(config.heads()))
		{
			testForEmergencyStop();
			auto fluxmap = readFluxmap(fluxsource, cylinder, head);
			fluxsink.writeFlux(cylinder, head, *fluxmap);
		}
    }
}

File: java/com/cowlark/fluxengine/algorithms/Reader.java
package com.cowlark.fluxengine.algorithms;

import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.LogMessage.BeginOperationLogMessage;
import com.cowlark.fluxengine.core.LogMessage.BeginReadOperationLogMessage;
import com.cowlark.fluxengine.core.LogMessage.EndOperationLogMessage;
import com.cowlark.fluxengine.core.LogMessage.EndReadOperationLogMessage;
import com.cowlark.fluxengine.core.LogMessage.OperationProgressLogMessage;
import com.cowlark.fluxengine.core.Logger;
import com.cowlark.fluxengine.core.Utils;
import com.cowlark.fluxengine.data.CylinderHead;
import com.cowlark.fluxengine.data.Disk;
import com.cowlark.fluxengine.data.DiskLayout;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.data.Image;
import com.cowlark.fluxengine.data.LogicalLocation;
import com.cowlark.fluxengine.data.LogicalTrackLayout;
import com.cowlark.fluxengine.data.PhysicalTrackLayout;
import com.cowlark.fluxengine.data.Sector;
import com.cowlark.fluxengine.data.Track;
import com.cowlark.fluxengine.decoders.Decoder;
import com.cowlark.fluxengine.fluxsink.FluxSink;
import com.cowlark.fluxengine.fluxsink.FluxSinkFactory;
import com.cowlark.fluxengine.fluxsource.FluxSource;
import com.cowlark.fluxengine.fluxsource.FluxSourceIterator;
import com.cowlark.fluxengine.imagewriter.ImageWriter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Disk read/write algorithms, ported from lib/algorithms/readerwriter.cc.
 */
public final class Reader
{
    static enum ReadResult
    {
        GOOD_READ, BAD_AND_CAN_RETRY, BAD_AND_CAN_NOT_RETRY
    }

    static enum BadSectorsState
    {
        HAS_NO_BAD_SECTORS, HAS_BAD_SECTORS
    }

    private Reader()
    {
    }

    public static void readDiskCommand(ConfigProto config,
                                       DiskLayout diskLayout,
                                       FluxSource fluxSource,
                                       Decoder decoder,
                                       Disk disk)
    {
        FluxSinkFactory outputFluxSinkFactory = null;
        if (config.getDecoder().hasCopyFluxTo())
            outputFluxSinkFactory =
                    FluxSinkFactory.create(config, config.getDecoder().getCopyFluxTo());

        Map<CylinderHead, List<Track>> tracksByLogicalLocation = new HashMap<>();
        for (Map.Entry<CylinderHead, Track> entry : disk.tracksByPhysicalLocation.entries())
        {
            Track track = entry.getValue();
            tracksByLogicalLocation.computeIfAbsent(
                    new CylinderHead(track.ltl.logicalCylinder, track.ltl.logicalHead),
                    k -> new ArrayList<>()).add(track);
        }

        Logger.log(new BeginOperationLogMessage("Reading and decoding disk"));

        if (fluxSource.isHardware())
            disk.rotationalPeriod = Common.measureDiskRotation(config);
        else
            disk.rotationalPeriod = Common.getRotationalPeriodFromConfig(config);

        try (FluxSink outputFluxSink = outputFluxSinkFactory != null ?
                outputFluxSinkFactory.create() :
                null)
        {
            int index = 0;
            for (Map.Entry<CylinderHead, LogicalTrackLayout> entry :
                    diskLayout.layoutByLogicalLocation.entrySet())
            {
                CylinderHead logicalLocation = entry.getKey();
                LogicalTrackLayout ltl = entry.getValue();
                Logger.log(new OperationProgressLogMessage(
                        index * 100 / diskLayout.layoutByLogicalLocation.size()));
                index++;

                Common.testForEmergencyStop();

                List<Track> trackFluxes = tracksByLogicalLocation.computeIfAbsent(
                        logicalLocation,
                        k -> new ArrayList<>());
                List<Sector> trackSectors = new ArrayList<>();
                readAndDecodeTrack(
                        config,
                        diskLayout,
                        fluxSource,
                        decoder,
                        ltl,
                        trackFluxes,
                        trackSectors);

                /* Replace all tracks on the disk by the new combined set. */

                for (Track flux : trackFluxes)
                    disk.tracksByPhysicalLocation.removeAll(new CylinderHead(
                            flux.ptl.physicalCylinder,
                            flux.ptl.physicalHead));
                for (Track flux : trackFluxes)
                    disk.tracksByPhysicalLocation.put(
                            new CylinderHead(
                                    flux.ptl.physicalCylinder,
                                    flux.ptl.physicalHead),
                            flux);

                /* Likewise for sectors. */

                for (Sector sector : trackSectors)
                    disk.sectorsByPhysicalLocation.removeAll(sector.physicalLocation);
                for (Sector sector : trackSectors)
                    disk.sectorsByPhysicalLocation.put(sector.physicalLocation, sector);

                if (outputFluxSink != null)
                {
                    for (Track data : trackFluxes)
                        outputFluxSink.addFlux(
                                data.ptl.physicalCylinder,
                                data.ptl.physicalHead,
                                data.fluxmap);
                }

                if (config.getDecoder().getDumpRecords())
                {
                    List<com.cowlark.fluxengine.data.Record> sortedRecords = new ArrayList<>();
                    for (Track data : trackFluxes)
                        sortedRecords.addAll(data.records);
                    sortedRecords.sort(Comparator.comparingDouble(r -> r.startTimeNs));

                    System.out.println("\nRaw (undecoded) records follow:\n");
                    for (com.cowlark.fluxengine.data.Record record : sortedRecords)
                    {
                        System.out.printf(
                                "I+%.2fus with %.2fus clock%n",
                                record.startTimeNs / 1000.0,
                                record.clockNs / 1000.0);
                        Utils.hexdump(System.out, record.rawData);
                        System.out.println();
                    }
                }

                if (config.getDecoder().getDumpSectors())
                {
                    List<Sector> sectors = collectSectors(trackSectors, false);
                    sectors.sort(Comparator.comparing((Sector s) -> s.location.logicalCylinder())
                            .thenComparing((Sector s) -> s.location.logicalHead())
                            .thenComparing((Sector s) -> s.location.logicalSector()));

                    System.out.println("\nDecoded sectors follow:\n");
                    for (Sector sector : sectors)
                    {
                        System.out.printf(
                                "%d.%02d.%02d: I+%.2fus with %.2fus clock: " + "status %s%n",
                                sector.location.logicalCylinder(),
                                sector.location.logicalHead(),
                                sector.location.logicalSector(),
                                sector.headerStartTimeNs / 1000.0,
                                sector.clockNs / 1000.0,
                                Sector.statusToString(sector.status));
                        Utils.hexdump(System.out, sector.data);
                        System.out.println();
                    }
                }

                /* track can't be modified below this point. */

                List<Sector> allSectors = new ArrayList<>();
                for (Sector sector : disk.sectorsByPhysicalLocation.values())
                    allSectors.add(sector);
                allSectors = collectSectors(allSectors);
                disk.image = new Image(allSectors);
            }
        }

        if (disk.image == null)
            disk.image = new Image();

        Logger.log(new EndOperationLogMessage("Read complete"));
    }

    public static void readDiskCommand(ConfigProto config,
                                       DiskLayout diskLayout,
                                       FluxSource fluxSource,
                                       Decoder decoder,
                                       ImageWriter writer)
    {
        Disk disk = new Disk();
        readDiskCommand(config, diskLayout, fluxSource, decoder, disk);

        writer.printMap(disk.image);
        if (config.getDecoder().hasWriteCsvTo())
            writer.writeCsv(disk.image, config.getDecoder().getWriteCsvTo());
        writer.writeImage(disk.image);
    }

    /* Given a set of sectors, deduplicates them sensibly (e.g. if there is a
     * good and bad version of the same sector, the bad version is dropped). */
    static List<Sector> collectSectors(List<Sector> trackSectors, boolean collapseConflicts)
    {
        Map<LogicalLocation, List<Sector>> sectors = new LinkedHashMap<>();
        for (Sector sector : trackSectors)
            sectors.computeIfAbsent(sector.location, k -> new ArrayList<>()).add(sector);

        List<Sector> sectorSet = new ArrayList<>();
        for (Map.Entry<LogicalLocation, List<Sector>> entry : sectors.entrySet())
        {
            List<Sector> bucket = entry.getValue();
            Sector newSector = bucket.get(0);
            for (int i = 1; i < bucket.size(); i++)
            {
                Sector right = bucket.get(i);
                if ((newSector.status == Sector.Status.OK) && (right.status == Sector.Status.OK) &&
                        (!newSector.data.equals(right.data)))
                {
                    if (!collapseConflicts)
                    {
                        Sector s = copySector(right);
                        s.status = Sector.Status.CONFLICT;
                        sectorSet.add(s);
                    }
                    Sector s = copySector(newSector);
                    s.status = Sector.Status.CONFLICT;
                    newSector = s;
                    continue;
                }
                if (newSector.status == Sector.Status.CONFLICT)
                    continue;
                if (right.status == Sector.Status.CONFLICT)
                {
                    newSector = right;
                    continue;
                }
                if (newSector.status == Sector.Status.OK)
                    continue;
                if (right.status == Sector.Status.OK)
                    newSector = right;
            }
            sectorSet.add(newSector);
        }

        return sectorSet;
    }

    static List<Sector> collectSectors(List<Sector> trackSectors)
    {
        return collectSectors(trackSectors, true);
    }

    private static Sector copySector(Sector sector)
    {
        Sector s = new Sector(sector.location);
        s.status = sector.status;
        s.position = sector.position;
        s.clockNs = sector.clockNs;
        s.headerStartTimeNs = sector.headerStartTimeNs;
        s.headerEndTimeNs = sector.headerEndTimeNs;
        s.dataStartTimeNs = sector.dataStartTimeNs;
        s.dataEndTimeNs = sector.dataEndTimeNs;
        s.physicalLocation = sector.physicalLocation;
        s.data = sector.data;
        s.records = sector.records;
        return s;
    }

    static class CombinationResult
    {
        BadSectorsState result;
        List<Sector> sectors;
    }

    static CombinationResult combineRecordAndSectors(List<Track> tracks,
                                                     Decoder decoder,
                                                     LogicalTrackLayout ltl)
    {
        CombinationResult cr = new CombinationResult();
        cr.result = BadSectorsState.HAS_NO_BAD_SECTORS;
        List<Sector> trackSectors = new ArrayList<>();

        /* Add the sectors which were there. */

        for (Track track : tracks)
            trackSectors.addAll(track.allSectors);

        /* Add the sectors which should be there. */

        for (int sectorId : ltl.diskSectorOrder)
        {
            Sector sector =
                    new Sector(new LogicalLocation(ltl.logicalCylinder, ltl.logicalHead, sectorId));

            sector.status = Sector.Status.MISSING;
            sector.physicalLocation = new CylinderHead(ltl.physicalCylinder, ltl.physicalHead);
            trackSectors.add(sector);
        }

        /* Deduplicate. */

        cr.sectors = collectSectors(trackSectors);
        if (cr.sectors.isEmpty())
            cr.result = BadSectorsState.HAS_BAD_SECTORS;
        for (Sector sector : cr.sectors)
            if (sector.status != Sector.Status.OK)
                cr.result = BadSectorsState.HAS_BAD_SECTORS;

        return cr;
    }

    static class ReadGroupResult
    {
        ReadResult result;
        List<Sector> combinedSectors;
    }

    static ReadGroupResult readGroup(DiskLayout diskLayout,
                                     Common.FluxSourceIteratorHolder fluxSourceIteratorHolder,
                                     LogicalTrackLayout ltl,
                                     List<Track> tracks,
                                     Decoder decoder,
                                     ConfigProto config)
    {
        ReadGroupResult rgr = new ReadGroupResult();
        rgr.result = ReadResult.BAD_AND_CAN_NOT_RETRY;

        /* Before doing the read, look to see if we already have the necessary
         * sectors. */

        {
            CombinationResult cr = combineRecordAndSectors(tracks, decoder, ltl);
            rgr.combinedSectors = cr.sectors;
            if (cr.result == BadSectorsState.HAS_NO_BAD_SECTORS)
            {
                /* We have all necessary sectors, so can stop here. */
                rgr.result = ReadResult.GOOD_READ;
                if (config.getDecoder().getSkipUnnecessaryTracks())
                    return rgr;
            }
        }

        for (int offset = 0; offset < ltl.groupSize; offset += diskLayout.headWidth)
        {
            int physicalCylinder = ltl.physicalCylinder + offset;
            int physicalHead = ltl.physicalHead;
            PhysicalTrackLayout ptl = diskLayout.layoutByPhysicalLocation.get(new CylinderHead(
                    physicalCylinder,
                    physicalHead));

            /* Do the physical read. */

            Logger.log(new BeginReadOperationLogMessage(physicalCylinder, physicalHead));

            FluxSourceIterator fluxSourceIterator =
                    fluxSourceIteratorHolder.getIterator(physicalCylinder, physicalHead);
            if (!fluxSourceIterator.hasNext())
                continue;

            Fluxmap fluxmap = fluxSourceIterator.next();
            Logger.log(new EndReadOperationLogMessage());
            Logger.logf("%d ms in %d bytes", (int) (fluxmap.duration() / 1e6), fluxmap.bytes());

            Track flux = decoder.decodeToSectors(fluxmap, ptl);
            flux.normalisedSectors = collectSectors(flux.allSectors);
            tracks.add(flux);

            /* Decode what we've got so far. */

            CombinationResult cr = combineRecordAndSectors(tracks, decoder, ltl);
            rgr.combinedSectors = cr.sectors;
            if (cr.result == BadSectorsState.HAS_NO_BAD_SECTORS)
            {
                /* We have all necessary sectors, so can stop here. */
                rgr.result = ReadResult.GOOD_READ;
                if (config.getDecoder().getSkipUnnecessaryTracks())
                    break;
            } else if (fluxSourceIterator.hasNext())
            {
                /* The flux source claims it can do more reads, so mark this
                 * group as being retryable. */
                rgr.result = ReadResult.BAD_AND_CAN_RETRY;
            }
        }

        return rgr;
    }

    private static void readAndDecodeTrack(ConfigProto config,
                                           DiskLayout diskLayout,
                                           FluxSource fluxSource,
                                           Decoder decoder,
                                           LogicalTrackLayout ltl,
                                           List<Track> tracks,
                                           List<Sector> combinedSectors)
    {
        if (fluxSource.isHardware())
            Common.measureDiskRotation(config);

        Common.FluxSourceIteratorHolder fluxSourceIteratorHolder =
                new Common.FluxSourceIteratorHolder(fluxSource);
        int retriesRemaining = config.getDecoder().getRetries();
        for (; ; )
        {
            ReadGroupResult rgr =
                    readGroup(diskLayout, fluxSourceIteratorHolder, ltl, tracks, decoder, config);
            combinedSectors.clear();
            combinedSectors.addAll(rgr.combinedSectors);
            if (rgr.result == ReadResult.GOOD_READ)
                break;
            if (rgr.result == ReadResult.BAD_AND_CAN_NOT_RETRY)
            {
                Logger.logf("no more data; giving up");
                break;
            }

            if (retriesRemaining == 0)
            {
                Logger.logf("giving up");
                break;
            }

            if (fluxSource.isHardware())
            {
                Common.adjustTrackOnError(fluxSource, ltl.physicalCylinder, config);
                Logger.logf("retrying; %d retries remaining", retriesRemaining);
                retriesRemaining--;
            }
        }
    }

}
